Support using scalar tensor for split

bpstark opened the pull request:
split requires an int input, however in tracing operators such as
size(axis) return a tensor, which is different behavior than when not
tracing. As such need to modify split to handle these cases.

Fixes #27551

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9b68afa7:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=b77c25dec056bf08b857f66b42ec6d2b21e547ff&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base b77c25de since Jan 22
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9b68afa7280213b4ec3a255a571289af46306195).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=89457464) from Jan 22



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

BowenBao commented on torch/tensor.py:
```diff
@@ -376,6 +376,12 @@ def split(self, split_size, dim=0):
+                split_size = int(split_size)
+                return super(Tensor, self).split(split_size, dim)
+            except ValueError:
+                return super(Tensor, self).split_with_sizes(split_size, dim)
```
Does `split_with_sizes` support `split_size` of type `Tensor`?

bpstark replied:
No it does not support a tensor as of now, but the idea is that if the tensor is not actually a scalar, it should fail in the same way that it currently does.

BowenBao reviewed (changes requested):
Please add test cases for the two new branches. Also is it possible for split_size to be passed in as a list of tensors?

bpstark commented:
@houseroad can you please pull in the latest changes.

thanks

Commit:
Support using scalar tensor for split

split requires an int input, however in tracing operators such as
size(axis) return a tensor, which is different behavior than when not
tracing. As such need to modify split to handle these cases.

bpstark commented:
rebase to deal with merge conflict

bpstark commented:
@houseroad Are there any issues with this change, or can it be merged?

houseroad reviewed (commented):
Could you ask @BowenBao @lara-hdr or @neginraoof to review this again?

lara-hdr commented on torch/tensor.py:
```diff
@@ -376,6 +376,12 @@ def split(self, split_size, dim=0):
         if isinstance(split_size, int):
             return super(Tensor, self).split(split_size, dim)
+        elif isinstance(split_size, Tensor):
+            try:
```
in this case it's a tensor of size 1?
Should we verify the size/dim instead of a try/except?

bpstark replied:
The problem is that size is returned as a torch.Size object which is a tensor in itself, so we cannot compare that to an int. Dim on the other hand does return an int, but only how many dims it has so it could be a tensor of dim 1, but that does not mean it is a scalar. In the end this was the only way I have found that can reliable work in all cases.

lara-hdr reviewed (approved):
Thanks for the fix!

lara-hdr commented:
@bpstark there is a test failure (seg fault), do you see the error locally or is it just on the CI?

bpstark commented:
@lara-hdr This appears to be a CI issue only, There should be nothing in my change that has any impact on the code that is seg faulting.

bpstark commented:
@houseroad I believe all should be good now.

houseroad reviewed (approved):
The failure is unrelated.
